### PR TITLE
[CMOB - 2174] Favourites

### PR DESCRIPTION
**Pull request by DavitTsiskarashvili**
https://spaceneobank.atlassian.net/browse/CMOB-2174

### CODE REVIEW

File: app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import androidx.room.Database
+import androidx.room.RoomDatabase
+import com.movies.data.local.dao.FavouriteMoviesDao
+import com.movies.data.local.entity.FavouriteMovieEntity
+
+@Database(entities = [FavouriteMovieEntity::class], version = 1)
+abstract class FavouriteMovieDatabase : RoomDatabase() {
+    abstract fun favouriteMovieDao(): FavouriteMoviesDao
+}
```
Comment (etozheluka) [inline app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt:11]: you can move version to constant, for example : 
```suggestion
@Database(entities = [FavouriteMovieEntity::class], version = DB_VERSION )
abstract class FavouriteMovieDatabase : RoomDatabase() {
    abstract fun favouriteMovieDao(): FavouriteMoviesDao
}
companion object {
        const val DB_VERSION = 1
    }
```

File: app/src/main/java/com/movies/data/local/module/DatabaseModule.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import android.content.Context
+import androidx.room.Room
+import com.movies.data.local.database.FavouriteMovieDatabase
+import org.koin.dsl.module
+
+private fun provideFavouriteMovieDatabase(context: Context): FavouriteMovieDatabase {
+    return Room.databaseBuilder(context, FavouriteMovieDatabase::class.java, "favourite_movie")
+        .fallbackToDestructiveMigration().build()
+}
```
Comment (etozheluka) [inline app/src/main/java/com/movies/data/local/module/DatabaseModule.kt:11]: also move db name to const

File: app/src/main/java/com/movies/data/repository/SearchRepositoryImpl.kt [hunk 3]
```diff
@@ -9,12 +10,16 @@ import com.movies.domain.repository.SearchRepository
...
-) : SearchRepository {
+    private val movieListDTOMapper: MovieListDTOMapper,
+    private val favouriteMoviesDao: FavouriteMoviesDao,
+    ) : SearchRepository {
     override suspend fun searchMovies(query: String): List<MovieDomainModel> {
         val remoteData = apiDataFetcher { fetchSearchedMovies.searchMovies(query) }
         if (remoteData is Resource.Success) {
-            return movieListDTOMapper(remoteData.data)
+            return movieListDTOMapper(remoteData.data).map {
+                it.isFavourite = favouriteMoviesDao.isFavouriteMovie(it.id)
```
Comment (etozheluka) [inline app/src/main/java/com/movies/data/repository/SearchRepositoryImpl.kt:20]: do you actually need this ? 
image

File: app/src/main/java/com/movies/domain/usecase/movies/GetMoviesUseCase.kt [hunk 4]
```diff
@@ -10,9 +10,9 @@ class GetMoviesUseCase(
 ) : BaseUseCase<CategoryType, List<MovieDomainModel>>() {
 
     override suspend fun invoke(params: CategoryType?): List<MovieDomainModel> {
-        return params?.let {
-            moviesRepository.fetchMovies(it)
-        } ?: emptyList()
+        return params.let {
+            moviesRepository.fetchMovies(it!!)
+        }
     }
```
Comment (etozheluka) [inline app/src/main/java/com/movies/domain/usecase/movies/GetMoviesUseCase.kt:17]: empty

File: app/src/main/java/com/movies/presentation/base/adapter/BaseAdapter.kt [hunk 5]
```diff
@@ -15,10 +15,13 @@ abstract class BaseAdapter<MODEL : Any> :
 
     abstract class BaseViewHolder<MODEL : Any>(binding: ViewBinding) :
         RecyclerView.ViewHolder(binding.root) {
-        abstract fun onBind(item: MODEL, onClickCallback: ((MODEL) -> Unit)?)
+        open fun onBind(item: MODEL, onClickCallback: ((MODEL) -> Unit)?){
+
+        }
```
Comment (etozheluka) [inline app/src/main/java/com/movies/presentation/base/adapter/BaseAdapter.kt]: you can fetch this to 1 line ->  open fun onBind(item: MODEL, onClickCallback: ((MODEL) -> Unit)?){ }

File: app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt [hunk 6]
```diff
@@ -22,19 +22,27 @@ class MovieAdapter : BaseAdapter<MovieUIModel>() {
...
                     posterImageView.loadImage(poster)
                     titleTextView.text = title
                     releaseYearTextView.text = releaseDate
                     ratingTextView.text = rating.toString()
+                    favouritesToggleButton.isChecked = isFavourite
 
                     root.setOnClickListener {
                         onClickCallback?.invoke(item)
                     }
+                    binding.favouritesToggleButton.setOnClickListener {
```
Comment (etozheluka) [inline app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt:43]: redundant binding.

File: app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt [hunk 7]
```diff
@@ -22,19 +22,27 @@ class MovieAdapter : BaseAdapter<MovieUIModel>() {
...
 
-    class MoviesViewHolder(private val binding: MovieItemBinding) :
-        BaseViewHolder<MovieUIModel>(binding) {
-        override fun onBind(item: MovieUIModel, onClickCallback: ((MovieUIModel) -> Unit)?) {
+    class MoviesViewHolder(val binding: MovieItemBinding) :
+        BaseMovieViewHolder<MovieUIModel>(binding) {
+        override fun onBindMovie(
+            item: MovieUIModel,
+            onClickCallback: ((MovieUIModel) -> Unit)?,
+            onFavouriteClickCallback: ((MovieUIModel, Boolean) -> Unit)?,
```
Comment (etozheluka) [inline app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt]: onFavouriteClickCallback -> onFavouriteClick

File: app/src/main/java/com/movies/presentation/base/adapter/BaseMovieAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package com.movies.presentation.base.adapter
+
+import androidx.viewbinding.ViewBinding
+
+abstract class BaseMovieAdapter<MODEL : Any, VH : BaseMovieAdapter.BaseMovieViewHolder<MODEL>> :
+    BaseAdapter<MODEL, VH>() {
+
+    private var onFavouriteClickCallback: ((MODEL, Boolean) -> Unit)? = null
```
Comment (etozheluka) [inline app/src/main/java/com/movies/presentation/base/adapter/BaseMovieAdapter.kt]: onFavouriteClickCallback -> onFavouriteClick

File: build.gradle [hunk 9]
```diff
@@ -2,5 +2,5 @@
 plugins {
     id 'com.android.application' version '8.0.2' apply false
     id 'com.android.library' version '8.0.2' apply false
-    id 'org.jetbrains.kotlin.android' version '1.8.20' apply false
+    id 'org.jetbrains.kotlin.android' version '1.7.10' apply false
```
Comment (etozheluka) [inline build.gradle:5]: ეს როგორც მახსოვს ტყუილად გავაკეთეთ

File: app/src/main/java/com/movies/data/repository/SearchRepositoryImpl.kt (hunk 3 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/repository/SearchRepositoryImpl.kt:20]: that is an optional search design. See the one on the left of that version, it has favourites too

File: app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt (hunk 1 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt:11]: I prefer in this way :3

File: app/src/main/java/com/movies/data/local/module/DatabaseModule.kt (hunk 2 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/local/module/DatabaseModule.kt:11]: same here

File: app/src/main/java/com/movies/domain/usecase/movies/GetMoviesUseCase.kt (hunk 4 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/domain/usecase/movies/GetMoviesUseCase.kt:17]: there should be, it makes more readable where the class end

File: app/src/main/java/com/movies/presentation/base/adapter/BaseAdapter.kt (hunk 5 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/base/adapter/BaseAdapter.kt]: done

File: app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt (hunk 6 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt:43]: already done on details pr

File: app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt (hunk 7 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/home/adapter/MovieAdapter.kt]: done

File: app/src/main/java/com/movies/presentation/base/adapter/BaseMovieAdapter.kt (hunk 8 above)
Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/base/adapter/BaseMovieAdapter.kt]: done

File: app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt (hunk 1 above)
Comment (etozheluka) [reply app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt:11]: why then you use constants for base url ?

Comment (etozheluka) [reply app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt:11]: you can paste whole string 😐

Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/data/local/database/FavouriteMovieDatabase.kt:11]: it's ugly there but not here

File: app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt [hunk 10]
```diff
@@ -76,14 +92,55 @@ class HomeFragment : BaseFragment<HomeViewModel>() {
...
+            viewModel.searchMovies(query = searchInput)
+        } else {
+            handleData(true)
+            viewModel.getMovies()
+        }
+    }
+
+    private fun handleFavouriteButton() {
+        movieAdapter.onFavouriteClickListener { favouriteMovie, isClicked ->
+            if (isClicked) {
```
Comment (AnaRusadze) [inline app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt]: isClicked მგონი ცუდი სახელია, წაშლის შემთხვევაშიც clicked ექშენი კეთდება. isSelected ხო არ ჯობია?

Comment (DavitTsiskarashvili) [reply app/src/main/java/com/movies/presentation/home/ui/HomeFragment.kt]: changed name into isChecked

**Review by etozheluka — APPROVED** [review]

**Review by AnaRusadze — APPROVED** [review]

**Review by AleksandreGalogreBaramidze — APPROVED** [review]

**Pull request merged**